ReadMe.txt
    BEDBUG - A family of FLEA-inspired CSPRNGs and Stream Ciphers
    BEDBUG is a BEDBUG with a 128-word, 256-word or 512-word internal state array
    BEDBUG may be seeded with a 4096-bit, an 8192-bit or a 16384-bit key
    BEDBUG is Copyright C.C.Kayne 2014, GNU GPL V.3, cckayne@gmail.com
    BEDBUG is based on FLEA and other PRNG insights by Bob Jenkins. Public Domain.

What is BEDBUG?

BEDBUG is a small, fast, cryptographically secure pseudo-random number generator (CSPRNG) and stream cipher. It exhibits uniform distribution, mixes rapidly (with worst-case avalanche better than 16-bits), has no detected bias, and comes in three variants: BEDBUG128, with an internal state array of 128+3 32-bit words; BEDBUG256, with an internal state of 256+3 words; and BEDBUG512 with a 512+3-word state. The former permit seeding with a key of up to 4096 or 8192 bits, the latter with a 16384-bit key.

The BEDBUG generator itself is a mere 5 lines of code:

		e = state[d % STATESIZE] - rot(b,iii);
		state[d % STATESIZE] = b ^ rot(c,jjj);
		b = c + rot(d,kkk);
		c = d + e;
		d = e + state[b % STATESIZE];
		
Three rotations, two pseudo-random lookups. 

The default BEDBUG configuration (B-RSW) alternates the rotation constants unpredictably. These four sets of three values differ between BEDBUG128, BEDBUG256 and BEDBUG512 and were selected, tuned and tested in each case for optimal avalanche, which is never less than 16.5 bits. The quantities are:

for BEDBUG128 -

1. iii = 14; jjj =  9; kkk = 16; // avalanche: 17.31 bits (worst case)
2. iii = 17; jjj =  8; kkk = 14; // avalanche: 17.31 bits (worst case)
3. iii = 17; jjj = 19; kkk = 30; // avalanche: 17.19 bits (worst case)
4. iii = 15; jjj =  9; kkk = 24; // avalanche: 17.19 bits (worst case)

for BEDBUG256 -

1. iii = 30; jjj = 13; kkk = 24; // avalanche: 17.00 bits (worst case)
2. iii =  5; jjj = 19; kkk = 20; // avalanche: 16.94 bits (worst case)
3. iii = 21; jjj =  4; kkk = 10; // avalanche: 16.84 bits (worst case)
4. iii = 29; jjj = 15; kkk =  6; // avalanche: 16.72 bits (worst case)

for BEDBUG512 -

1. iii = 17; jjj =  3; kkk =  4; // avalanche: 16.56 bits (worst case)
2. iii = 15; jjj = 18; kkk = 23; // avalanche: 16.52 bits (worst case)
3. iii =  3; jjj = 23; kkk = 24; // avalanche: 16.52 bits (worst case)
4. iii =  3; jjj = 15; kkk =  8; // avalanche: 16.53 bits (worst case)

The three BEDBUG variants presented here have been implemented identically in C and Pascal and can be found in their respective directories along with b, bb and bbb - three short programs to check their output. BEDBUG was developed and tested with the Tiny C compiler (tcc) and Free Pascal (fpc) under Win32. The NIST statistical tests were run under Linux.

To view BEDBUG's abbreviated internal state, please uncomment "#define TEST" in the C files and "{$define TEST}" in the Pascal files and "#define VERBOSE" and "{$define VERBOSE}" to display the full result & state arrays as they are generated on each call to bbXXX.

PRNG tests passed by BEDBUG:

* NIST suite of statistical tests for randomness: all tests for all seeds
* George Marsaglia's DIEHARD suite: all tests for all seeds
* Bob Jenkins' countx for sub-sequences of length 2-12 (2**35 trials)
* Bob Jenkins' rngav (avalanche better than 17.0 bits average)
* John Walker's ENT test for entropy
* C.C.Kayne's GapTrial (value-spacings normal over all modulos)
* C.C.Kayne's ModTrial (value-distributions uniform over all modulos)
* C.C.Kayne's SeqTrial (same-value sequences normal over all modulos)
* C.C.Kayne's PopTrial (bit-populations iid-normal over all trial ranges)
* C.C.Kayne's Visual 1 & 2

(Please view the Results.txt files in the relevant folders or verify the recorded results by re-running the tests themselves. The principal BEDBUG repositories are listed at the foot of this document.)


Why BEDBUG?

Because there are not yet enough NSA- and NIST-free strong light ciphers in the world. ISAAC is one to rule them all, FLEA came and went briefly, and now hops BEDBUG into their traces, thanks to Bob Jenkins' gratifyingly published insights into the art of creating the PRNG.


Why "bedbug"?

Because it shares kinship and vocation with "flea", is tiny and mean, and is doubtless destined to irritate the bejaysus out of crackers and cryptanalysts.

Have fun with BEDBUG, and don't hesitate to get in touch if you encounter issues with any of the files included in this distribution. As always, your comments, suggestions, criticism and general feedback are welcomed.

C.C.Kayne
cckayne@gmail.com
https://github.com/cckayne/BEDBUG
https://code.google.com/p/bedbug-csprng-stream-cipher/
https://sourceforge.net/projects/bedbug-csprng-stream-cipher/
https://sourceforge.net/projects/motet-cipher-scrambler/
https://github.com/cckayne/MOTET
https://github.com/cckayne/AUM


ChangeLog.txt
07eea3c (HEAD, BEDBUG-SourceForge/master, BEDBUG-Github/master, master) BEDBUG CSPRNG & Tests, version 1.3.1 Docs augmented with "On Using PRNGs as the Basis for Stream Ciphers" SamTrial added to Tests.
dc14cb4 (BEDBUG-GoogleCode/master) BEDBUG CSPRNG & Tests, version 1.2.0 Changelog updated.
d30680e BEDBUG CSPRNG & Tests, version 1.3.0 countx test enhanced and given its own folder.
09310db !!
b70e23f !
892c44c BEDBUG CSPRNG & Tests, version 1.2.0 countx test enhanced and given its own folder.
58cee24 BEDBUG CSPRNG & Tests, version 1.2.0 Safeties added to CSPRNG seeding functions: keys are truncated if too long.
269825a BEDBUG CSPRNG & Tests, version 1.1.0 Minor text and code edits.
044c6a7 BEDBUG CSPRNG & Tests, version 1.1.0
68824b9 BEDBUG CSPRNG & Tests, version 1.1.0 Minor ReadMe edits
c6b7293 BEDBUG CSPRNG & Tests, version 1.1.0 Inclusion of MOTET Minor text & comment edits
c443630 BEDBUG CSPRNG & Tests, version 1.1.0
044a87b BEDBUG CSPRNG & Tests, version 1.1.0 MOTET Cipher & SE Scrambler application added. Various optimizations.
ae6f825 Minor edits.
caf3a03 Minor update to die.pas and text edits.
1f06e62 Minor edits and tidying.
afb1c1f Initial commit - BEDBUG CSPRNG with TESTS.


C/bb256.c
/*  BEDBUG256 - A FLEA-inspired CSPRNG and Stream Cipher
    BEDBUG256 is a BEDBUG with a 256-byte internal state array
    BEDBUG256 may be seeded with a 8192-bit key
    BEDBUG256 Copyright C.C.Kayne 2014, GNU GPL V.3, cckayne@gmail.com
    BEDBUG256 is based on FLEA and other PRNG insights by Bob Jenkins. Public Domain.
*/
#include <stdio.h>
#include <stddef.h>
#include <string.h>
#include "bb256.h"

/* are we testing? */
//#define TEST
/* verbose test output */
//#define VERBOSE
/* byte size of state array */
#define STSZ 256
#define STM1 STSZ-1
#define STBYTES STSZ*4
#define STBITS 128+STBYTES*8
/* BB256b performs an extra pseudo-random lookup */
#define B
/* BB256 option: switch ROT constants at each bb256() call */
#define RSW
/* 2**32/phi, where phi is the golden ratio */
#define GOLDEN 0x9e3779b9

// BB256 STATE
static ub4 rsl[STSZ], state[STSZ];
static ub4 b,c,d,e, rcnt=0;

// BB512 ROT switcher
typedef struct Rsw { ub4 iii; ub4 jjj; ub4 kkk; };
static ub4 ri=0;
static const struct Rsw rsw[4] = {
#ifdef B
30, 13, 24, // avalanche: 17.00 bits (worst case)
 5, 19, 20, // avalanche: 16.94 bits (worst case)
21,  4, 10, // avalanche: 16.84 bits (worst case)
29, 15,  6  // avalanche: 16.72 bits (worst case)
#else
20, 14,  8, // avalanche: 17.00 bits (worst case)
20,  2, 22, // avalanche: 16.97 bits (worst case)
22, 23, 24, // avalanche: 16.78 bits (worst case)
 7,  8, 22  // avalanche: 16.72 bits (worst case)
#endif
};

#define rot(x,k) ((x<<k)|(x>>(32-k)))


#ifdef TEST
static ub4 bcnt=0;
static void statepeek(void) {
	register ub4 i, gcnt=0;
	++bcnt;
	printf("%3u) bb256 using rsw[%1u]...\n",bcnt,ri);
	for (i=0; i<STSZ; i++) {
		#ifdef VERBOSE
		printf("rsl %3u: %11u %c %02X  | state %3u: %11u %c %02X\n",
			i,rsl[i],rsl[i] % 26 + 'A',rsl[i] & 255,
			i,state[i],state[i] % 26 + 'A',state[i] & 255);
		#endif
		if (state[i]==GOLDEN) ++gcnt;
	}
	printf("     b = %11u %c %02X\n     c = %11u %c %02X\n     d = %11u %c %02X\n",
		b,b % 26+'A',b & 255,c,c % 26+'A',c & 255, d,d % 26+'A',d & 255);
	printf("     %u GOLDEN\n",gcnt);
}
#endif


// BEDBUG256 is filled every 256 rounds
static void bb256(void) {
    register ub4 i;
	for (i=0; i<STSZ; i++) {
		e = state[d & STM1] - rot(b,rsw[ri].iii);
		state[d & STM1] = b ^ rot(c,rsw[ri].jjj);
		b = c + rot(d,rsw[ri].kkk);
		c = d + e;
		#ifdef B
		d = e + state[b & STM1];
		#else
		d = e + state[i];
		#endif
		rsl[i] = d;
	}
	#ifdef TEST
	statepeek();
	#endif
	#ifdef RSW
	ri = (c & 3);
	#endif
}


// reset/initialize BEDBUG256 (obligatory)	
void bb256_Reset(void) {
	register ub4 i;
	rcnt=0; ri = 0;
	b = c = d = e = GOLDEN;
	for (i=0; i<STSZ; i++) {
		state[i] = GOLDEN; rsl[i] = 0;
	}
}


// obtain a BEDBUG pseudo-random value in [0..2**32]
ub4 bb256_Random(void) {
	ub4 r = rsl[rcnt];
	++rcnt;
	if (rcnt==STSZ) {
		bb256();
		rcnt = 0;
	}
	return r;
}


// seed BEDBUG256 with a 8192-bit block of 4-byte words (Bob Jenkins method) 
void bb256_SeedW(char *seed, int rounds)
{
	register ub4 i,l;
	char s[STBYTES*2];
	l=strlen(seed);
	if (l>STBYTES) l=STBYTES;
	memset(s,0,l+1);
	/* truncate seed to state-size if necessary */
	for (i=0; i<l; i++) s[i] = seed[i];
	bb256_Reset();
	memcpy((char *)state, (char *)s, l);
	bb256();
	for (i=0; i<rounds; i++) bb256_Random();  
}
